src/aspects/ec2-automated-shutdown/index.ts

```diff
@@ -0,0 +1,178 @@
+            const lambda = new SecureFunction(stack, 'Ec2ShutdownFunction', {
+                code: Code.fromAsset(join(__dirname, 'handler')),
+                runtime: Runtime.NODEJS_20_X,
+                handler: 'index.handler',
+                timeout: Duration.seconds(15),
+                encryption: this.props.encryption,
+                ...this.props.lambdaConfiguration
+            });
+
+            if (!Ec2AutomatedShutdown.policyByStack.has(stackId)) {
+                const policy = new PolicyStatement({
+                    actions: ['ec2:StopInstances'],
+                    effect: Effect.ALLOW
+                });
+                Ec2AutomatedShutdown.policyByStack.set(stackId, policy);
+                lambda.role.addToPolicy(policy);
+            }
+            Ec2AutomatedShutdown.lambdaByStack.set(stackId, lambda);
+        }
+        return Ec2AutomatedShutdown.lambdaByStack.get(stackId)!;
+    }
+
+    private addInstanceToPolicy(
+        instance: Instance | CfnInstance,
+        stack: Stack
+    ): void {
+        const stackId = stack.stackId;
+        const policy = Ec2AutomatedShutdown.policyByStack.get(stackId);
+
+        if (!policy) {
+            throw new Error('Policy not initialized');
+        }
+
+        const instanceArn = `arn:${Stack.of(instance).partition}:ec2:${stack.region}:${stack.account}:instance/${
+            instance instanceof Instance ? instance.instanceId : instance.ref
+        }`;
+
+        policy.addResources(instanceArn);
+    }
+
+    private createShutdownMechanism(instance: IConstruct, stack: Stack): void {
+        const lambdaFunction = this.getLambdaFunction(stack);
+
+        let instanceId: string;
+        if (instance instanceof Instance) {
+            instanceId = instance.instanceId;
+        } else if (instance instanceof CfnInstance) {
+            instanceId = instance.ref;
+        } else {
+            throw new Error('Unsupported instance type');
+        }
+
+        this.addInstanceToPolicy(instance, stack);
+
+        const topic = new Topic(stack, `${instance.node.id}Ec2ShutdownTopic`, {
+            displayName: `Ec2 Shutdown Topic for ${instance.node.id}`,
+            masterKey: this.props.encryption
+        });
+
+        lambdaFunction.function.addEventSource(new SnsEventSource(topic));
+
+        topic.addToResourcePolicy(
+            new PolicyStatement({
+                sid: 'AllowPublishThroughSSLOnly',
+                effect: Effect.DENY,
+                principals: [new AnyPrincipal()],
+                actions: ['sns:Publish'],
+                resources: [topic.topicArn],
+                conditions: {
+                    Bool: {
+                        'aws:SecureTransport': 'false'
+                    }
+                }
+            })
+        );
+
+        const metricConfig =
+            this.props.metricConfig || this.defaultMetricConfig;
+
+        const metric = new Metric({
+            namespace: metricConfig.namespace,
```

Comment:
The metric namespace should always be "AWS/EC2" and can be hardcoded here unless I'm mistaken?